File: core/src/main/java/com/terralite/core/registry/SimpleMutableRegistry.java
package com.terralite.core.registry;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class SimpleMutableRegistry<T> implements MutableRegistry<T> {
    private final RegistryKey<T> key;
    private final Map<ResourceId, T> entries = new LinkedHashMap<>();
    private boolean frozen;

    public SimpleMutableRegistry(RegistryKey<T> key) {
        this.key = Objects.requireNonNull(key, "key");
    }

    @Override
    public RegistryKey<T> key() {
        return key;
    }

    @Override
    public T register(ResourceId id, T value) {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(value, "value");

        if (frozen) {
            throw new IllegalStateException("Registry is frozen: " + key.id());
        }

        if (entries.containsKey(id)) {
            throw new IllegalArgumentException("Duplicate registry entry: " + id);
        }

        entries.put(id, value);
        return value;
    }

    @Override
    public T replace(ResourceId id, T value) {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(value, "value");

        if (frozen) {
            throw new IllegalStateException("Registry is frozen: " + key.id());
        }

        if (!entries.containsKey(id)) {
            throw new IllegalArgumentException("No registry entry to replace: " + id);
        }

        entries.put(id, value);
        return value;
    }

    @Override
    public Optional<T> get(ResourceId id) {
        return Optional.ofNullable(entries.get(id));
    }

    @Override
    public T require(ResourceId id) {
        T value = entries.get(id);
        if (value == null) {
            throw new IllegalArgumentException("Missing registry entry: " + id + " in " + key.id());
        }
        return value;
    }

    @Override
    public boolean contains(ResourceId id) {
        return entries.containsKey(id);
    }

    @Override
    public Collection<ResourceId> ids() {
        return Collections.unmodifiableSet(entries.keySet());
    }

    @Override
    public Collection<T> values() {
        return Collections.unmodifiableCollection(entries.values());
    }

    @Override
    public int size() {
        return entries.size();
    }

    @Override
    public boolean isFrozen() {
        return frozen;
    }

    @Override
    public FrozenRegistry<T> freeze() {
        frozen = true;
        return new SimpleFrozenRegistry<>(key, entries);
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/SimpleMutableRegistry.kt
package com.terralite.core.registry

/** Default [MutableRegistry] backed by a linked hash map (insertion order preserved). */
class SimpleMutableRegistry<T>(private val registryKey: RegistryKey<T>) : MutableRegistry<T> {

    private val entries: LinkedHashMap<ResourceId, T> = LinkedHashMap()
    private var frozen = false

    override fun key(): RegistryKey<T> = registryKey

    override fun register(id: ResourceId, value: T): T {
        check(!frozen) { "Registry is frozen: ${registryKey.id}" }
        require(!entries.containsKey(id)) { "Duplicate registry entry: $id" }
        entries[id] = value
        return value
    }

    override fun replace(id: ResourceId, value: T): T {
        check(!frozen) { "Registry is frozen: ${registryKey.id}" }
        require(entries.containsKey(id)) { "No registry entry to replace: $id" }
        entries[id] = value
        return value
    }

    override fun get(id: ResourceId): T? = entries[id]

    override fun require(id: ResourceId): T =
        entries[id] ?: throw IllegalArgumentException("Missing registry entry: $id in ${registryKey.id}")

    override fun contains(id: ResourceId): Boolean = entries.containsKey(id)

    override fun ids(): Collection<ResourceId> = entries.keys.toSet()

    override fun values(): Collection<T> = entries.values.toList()

    override fun size(): Int = entries.size

    override fun isFrozen(): Boolean = frozen

    override fun freeze(): FrozenRegistry<T> {
        frozen = true
        return SimpleFrozenRegistry(registryKey, entries)
    }
}
